browser/src/react-redux/App/Signup/SignupForm.jsx

```diff
@@ -0,0 +1,156 @@
+import React, { Component } from "react";
+import axios from "axios";
+import Form from "./Form";
+import utility from "./signupUtility";
+
+class signupForm extends Component {
+  constructor() {
+    super();
+    this.state = {
+      email: "",
+      password: "",
+      confirmPassword: "",
+      errors: {
+        confirmPassword: "",
+        password: "",
+        email: "",
+        general: ""
+      },
+      buttonDisabled: false
+    };
+    this.handleChange = this.handleChange.bind(this);
+    this.onSubmit = this.onSubmit.bind(this);
+  }
+
+  handleChange(key, e) {
+    this.setState({
+      [key]: e
+    });
+  }
+  onSubmit(e) {
+    this.setState({
+      errors: {
+        general: "",
+        email: "",
+        password: ""
+      }
+    });
+
+    e.preventDefault();
+
+    const user = {
+      email: this.state.email,
+      password: this.state.password
+    };
+
+    axios
+      .post("/api/user", user)
+      .then(res => {
+        if (res.request.status === 201) {
+          console.log(`Success!: ${JSON.stringify(res.data.user, null, "  ")}`);
+        }
+      })
+      .catch(err => {
+        console.log(
+          `Error!: ${JSON.stringify(err.response.data.error, null, "  ")}`
+        );
+        if (err.response.data.error.errors) {
+          let newErrors = utility.extractErrors(err.response.data.error.errors);
+
+          this.setState({
+            errors: { ...this.state.errors, ...newErrors }
+          });
+        } else {
+          this.setState({
+            errors: {
+              general: err.response.data.error.message
+            }
+          });
+        }
+      });
+  }
+  componentDidUpdate(prevProps, prevState) {
+    //check if password and confirmPassword fields match. If not, set error message to state.
+
+    if (
+      this.state.confirmPassword &&
+      this.state.password != this.state.confirmPassword &&
+      !this.state.errors.confirmPassword
+    ) {
+      this.setState({ errors: { confirmPassword: "Passwords do not match." } });
+    } else if (prevState.errors.confirmPassword) {
+      this.setState({
+        errors: {
+          confirmPassword: ""
+        }
+      });
+    }
+    //disable submit until following requirements are met
```

Comment:
Instead of having state for the button disabled && clouding up `componentDidUpdate`, we could do something like this in the render:

```
    const disableButton = 
      !this.state.email ||
      !this.state.password ||
      !this.state.confirmPassword ||
      (this.state.confirmPassword && this.state.password !== this.state.confirmPassword);
```

Then we could get rid of `buttonDisabled` in state all-together.

Reply:
You can even create a method for this and just invoke the method in the disable property.

Reply:
I support a method.  It allows if-else statements for readability as well as puts all of this disable conditions in one place. In this case, you can have the same logic but have it return of true or false instead of setting state.